File: modules/ducktests/tests/ignitetest/tests/thin_client_ssl_test.py
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
This module contains ssl tests.
"""
from ducktape.mark import matrix

from ignitetest.services.ignite import IgniteService
from ignitetest.services.ignite_app import IgniteApplicationService
from ignitetest.services.utils.control_utility import ControlUtility
from ignitetest.services.utils.ignite_configuration import IgniteConfiguration, IgniteThinClientConfiguration
from ignitetest.services.utils.ssl.client_connector_configuration import ClientConnectorConfiguration
from ignitetest.services.utils.ssl.connector_configuration import ConnectorConfiguration
from ignitetest.services.utils.ssl.ssl_params import SslParams, DEFAULT_SERVER_KEYSTORE, DEFAULT_CLIENT_KEYSTORE, \
    DEFAULT_ADMIN_KEYSTORE
from ignitetest.utils import ignite_versions, cluster
from ignitetest.utils.ignite_test import IgniteTest
from ignitetest.utils.version import IgniteVersion, DEV_BRANCH, LATEST


# pylint: disable=W0223
class ThinClientSslTest(IgniteTest):
    """
    cluster - cluster size.
    JAVA_CLIENT_CLASS_NAME - running classname.
    """

    JAVA_CLIENT_CLASS_NAME = "org.apache.ignite.internal.ducktest.tests.thin_client_test.ThinClientSelfTestApplication"

    @cluster(num_nodes=2)
    @matrix(server_version=[str(DEV_BRANCH), str(LATEST)], thin_client_version=[str(DEV_BRANCH), str(LATEST)])
    def test_ssl_connection(self, server_version, thin_client_version):
        """
        Thin client ssh connection test.
        """
        root_dir = self.test_context.globals.get("install_root", "/opt")

        server_ssl = SslParams(root_dir=root_dir, key_store_jks=DEFAULT_SERVER_KEYSTORE)

        server_config = IgniteConfiguration(
            version=IgniteVersion(server_version), ssl_params=server_ssl,
            connector_configuration=ConnectorConfiguration(ssl_enabled=True, ssl_params=server_ssl),
            client_connector_configuration=ClientConnectorConfiguration())

        ignite = IgniteService(self.test_context, server_config, 1)

        addresses = ignite.nodes[0].account.hostname + ":" + \
                    str(server_config.client_connector_configuration.port)

        thin_clients = IgniteApplicationService(
            self.test_context,
            IgniteThinClientConfiguration(addresses=addresses,
                                          version=IgniteVersion(thin_client_version),
                                          ssl_params=SslParams(root_dir=root_dir,
                                                               key_store_jks=
                                                               DEFAULT_CLIENT_KEYSTORE,
                                                               user='client',
                                                               password='123456')),
            java_class_name=self.JAVA_CLIENT_CLASS_NAME)

        ignite.start()
        thin_clients.run()
        ignite.stop()

File: modules/core/src/test/java/org/apache/ignite/ssl/ThinClientSSLTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.ignite.ssl;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import javax.cache.configuration.Factory;
import javax.net.ssl.SSLContext;
import org.apache.ignite.IgniteCache;
import org.apache.ignite.Ignition;
import org.apache.ignite.client.ClientCache;
import org.apache.ignite.client.ClientException;
import org.apache.ignite.client.IgniteClient;
import org.apache.ignite.client.SslMode;
import org.apache.ignite.configuration.ClientConfiguration;
import org.apache.ignite.configuration.ClientConnectorConfiguration;
import org.apache.ignite.configuration.IgniteConfiguration;
import org.apache.ignite.testframework.GridTestUtils;
import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
import org.junit.Test;

public class ThinClientSSLTest extends GridCommonAbstractTest {

    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
        IgniteConfiguration igniteCfg = super.getConfiguration(igniteInstanceName);

        igniteCfg.setSslContextFactory(serverSSLFactory());

        ClientConnectorConfiguration clientConnectorCfg = new ClientConnectorConfiguration()
                .setSslEnabled(true)
                .setSslClientAuth(true)
                .setUseIgniteSslContextFactory(false)
                .setSslContextFactory(clientConnectorSSLFactory());
        igniteCfg.setClientConnectorConfiguration(clientConnectorCfg);

        return igniteCfg;
    }

    /**
     * @return SSL context factory to use on server nodes for communication between nodes in a cluster.
     */
    private Factory<SSLContext> serverSSLFactory() {
        return GridTestUtils.sslTrustedFactory("server", "trustone");
    }

    /**
     * @return SSL context factory to use in client connectors.
     */
    private Factory<SSLContext> clientConnectorSSLFactory() {
        return GridTestUtils.sslTrustedFactory("thinServer", "trusttwo");
    }

    private ClientConfiguration clientConfigurationWithoutSslContextFactory(String addr) {
        return new ClientConfiguration().setAddresses(addr)
                .setSslClientCertificateKeyStorePath(GridTestUtils.keyStorePath("thinClient"))
                .setSslTrustCertificateKeyStorePath(GridTestUtils.keyStorePath("trustone"))
                .setSslClientCertificateKeyStorePassword(GridTestUtils.keyStorePassword())
                .setSslTrustCertificateKeyStorePassword(GridTestUtils.keyStorePassword())
                .setSslMode(SslMode.REQUIRED)
                .setSslTrustAll(true);
    }

    private ClientConfiguration clientConfigurationWithSslContextFactory(String addr) {
        return new ClientConfiguration().setAddresses(addr)
                .setSslContextFactory(SSLContextFactoryForTests.thinClientSSLFactoryWithWrongTrustCertificate())
                .setSslMode(SslMode.REQUIRED)
                .setSslTrustAll(true);
    }


    /** {@inheritDoc} */
    @Override protected void beforeTestsStarted() throws Exception {
        startGrids(1);
    }

    /**
     * Checks that thin clients with SSL enabled and SetSSLTrust(true) without set SslContextFactory can join the cluster.
     *
     * @throws Exception If failed.
     */
    @Test
    public void testSetSSLTrustAllTrueWithoutSslContextFactory() throws Exception {
        String cacheName = "thinClientCache";
        List<IgniteClient> clients = new ArrayList<>(1);

        try {
            IgniteClient client = Ignition.startClient(clientConfigurationWithoutSslContextFactory("127.0.0.1:10800"));

            Map<Integer, Integer> expMap = new HashMap<>();

            ClientCache<Integer, Integer> cache = client.getOrCreateCache(cacheName);

            cache.put(1, 1);
            expMap.put(1, 1);

            IgniteCache<Integer, Integer> cache1 = grid(0).cache(cacheName);

            assertCacheContent(expMap, cache1);
        }
        catch (ClientException ex) {
            ex.printStackTrace();

            fail("Failed to start thin Java clients: " + ex.getMessage());
        }
        finally {
            for (IgniteClient client : clients)
                client.close();

            IgniteCache cache = grid(0).cache(cacheName);

            if (cache != null)
                cache.destroy();
        }
    }


    /**
     * Checks that thin clients with SSL enabled and SetSSLTrust(true) with set SslContextFactory can join the cluster.
     *
     * @throws Exception If failed.
     */
    @Test
    public void testSetSSLTrustAllTrueWithSslContextFactory() throws Exception {
        String cacheName = "thinClientCache";
        List<IgniteClient> clients = new ArrayList<>(1);

        try {
            IgniteClient client = Ignition.startClient(clientConfigurationWithSslContextFactory("127.0.0.1:10800"));

            Map<Integer, Integer> expMap = new HashMap<>();

            ClientCache<Integer, Integer> cache = client.getOrCreateCache(cacheName);

            cache.put(1, 1);
            expMap.put(1, 1);

            IgniteCache<Integer, Integer> cache1 = grid(0).cache(cacheName);

            assertCacheContent(expMap, cache1);
        }
        catch (ClientException ex) {
            ex.printStackTrace();

            fail("Failed to start thin Java clients: " + ex.getMessage());
        }
        finally {
            for (IgniteClient client : clients)
                client.close();

            IgniteCache cache = grid(0).cache(cacheName);

            if (cache != null)
                cache.destroy();
        }
    }

    private void assertCacheContent(Map<Integer, Integer> exp, IgniteCache<Integer, Integer> cache) {
        assertEquals("Cache has an unexpected size.", exp.size(), cache.size());
        for (Map.Entry<Integer, Integer> e : exp.entrySet()) {
            int key = e.getKey();
            Integer expVal = e.getValue();
            Integer actVal = cache.get(key);

            assertEquals("Cache contains an unexpected value for a key=" + key, expVal, actVal);
        }
    }
}
